### PR TITLE
Redefine lr_scheduler behavior

### PR DESCRIPTION
**Pull request by wwwjn**
## Context

The current warmup-stable-decay lr_scheduler behavior is not intuitive. For example, in `debug_model.toml`, the configurattion is: 
```
[lr_scheduler]
warmup_steps = 2  # lr scheduler warm up, normally 20% of the train steps
decay_ratio = 0.8  # lr scheduler decay ratio, 80% of the train steps
decay_type = "linear"
lr_min = 0.0
```
So we are expected warmup_steps =2, decay steps=8, total steps=10.

And we got the curret learning rate (blue line) There are 2 issues:
1. The max learning rate is not reaching 1 (expected max lr = 1 since we are calculating ratio here). 
2. Intuitively, the user would expect to see learning rate increase by 0.5 (suppose max_lr = 1) each step during warm up stage, and decrease by 1/8 each step during decay stage. But in blue line, the lr is increasing by 1/3, and decreasing by 1/9, which is counter-intuitive. 

Thus we propose a standard lr_scheduler behavior, which aligns with user's intuitive and the meaning of the parameter names. (the red line)
learning_rate_schedule

## Standard lr_scheduler behavior
- Warm up stage:  LR increase by 1/{warmup_steps}
- Stable stage: Length of  stable stage = total_train_step + 1 - warmup_stage - decay_stage. We manually add one step to stable stage (so we have a fake step 11), which is preventing if decay is enabled the step 10 learning rate drops to 0. 
- Decay stage:   LR decrease by 1/{decay_steps}

### CODE REVIEW

File: torchtitan/components/lr_scheduler.py [hunk 1]
```diff
@@ -146,13 +155,13 @@ def linear_warmup_stable_decay(
             # linear warmup
             # 0-indexed step, hence + 1 adjustments
             current_step += 1
-            curr_adjustment = float(current_step / (warmup_steps + 1))
-        elif current_step < warmup_stable_steps:
+            curr_adjustment = float(current_step / warmup_steps)
+        elif current_step < warmup_stable_steps or decay_steps == 0:
```
Comment (tianyu-l) [inline torchtitan/components/lr_scheduler.py]: I think we won't hit `decay_steps == 0` in this `elif`. We can remove, and do assertion in the `else` branch below.

File: torchtitan/components/lr_scheduler.py [hunk 2]
```diff
@@ -146,13 +155,13 @@ def linear_warmup_stable_decay(
             # linear warmup
             # 0-indexed step, hence + 1 adjustments
             current_step += 1
-            curr_adjustment = float(current_step / (warmup_steps + 1))
-        elif current_step < warmup_stable_steps:
+            curr_adjustment = float(current_step / warmup_steps)
```
Comment (tianyu-l) [inline torchtitan/components/lr_scheduler.py:161]: add an assertion here

File: tests/unit_tests/test_lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+            expected_lr = 0.1 * factor
+            self.assertAlmostEqual(
+                self.optimizer.param_groups[0]["lr"],
+                expected_lr,
+                places=6,
+                msg=f"Step {i}: Expected LR {expected_lr}, got {self.optimizer.param_groups[0]['lr']}",
+            )
+            lr_scheduler.step()
+
+    def test_cosine_decay(self):
```
Comment (tianyu-l) [inline tests/unit_tests/test_lr_scheduler.py]: The cosine and sqrt tests are less helpful/necessary right now. We can add them later if needed.

**Review by tianyu-l — APPROVED** [review]
awesome! had some comments
